### DISCUSSION TITLE
PIO shift register data transfer timing.

### DISCUSSION THREAD

**Discussion by andrewmarles** [grblHAL/RP2040#6] [General]
Just wondering if people have done much testing on the shift register code?  Has anybody else run into any timing issues?

When I was testing my board in my 3 axis mill, I found that I was getting inconsistent steps.  Digging a bit deeper I found that the setup/hold requirement of the 74HC595D might not always be met:

![shift_reg_bad](https://user-images.githubusercontent.com/6061539/131445737-36725913-c9fc-4d93-a8ef-3b3d4fc6db75.png)

According to the datasheet, the HC595 needs the data clocked on the rising edge.  PCB skew in this design is ~17 ps between clock and data so I do not believe it is a layout issue.

![shift_reg_good](https://user-images.githubusercontent.com/6061539/131446113-7b6bf00c-33d8-4cad-b8cb-eaef26c58c37.png)

I added side-set delays to ensure that the clock and data are aligned correctly and slowed them down to ensure that the data transfer is robust.  This should still allow for step rates around 300 KHz which is faster than most external drivers can handle.  Since these delays are in the PIO they should not slow down the application.

**Comment by terjeio** [comment]
> Just wondering if people have done much testing on the shift register code? Has anybody else run into any timing issues?

I have only tested with a scope - and I do not think many are using this driver as there are currently no breakout boards out in the wild unless Phil has shipped some of the test boards he made. Do you have one or have you made your own?

What kind of issues have you seen?

> I added side-set delays to ensure that the clock and data are aligned correctly and slowed them down to ensure that the data transfer is robust. 

Great - can you make a PR?

For now I'll test the current code with Phils board and my [CNC machine simulator](https://github.com/terjeio/CNC_Machine_Simulator), I want to see if that would detect timing issues.

**Reply by andrewmarles** [reply]
Yes, I did my own RP2040 board.  Still testing before I post the design.  It uses the HC595D for step/dir and TCA9539 for aux IO.  I will post a PR for the TCA9539 driver shortly as it is a very small change to get the outputs working with your PCA9654E code as a starting point.

**Reply by terjeio** [reply]
Thanks for feedback.

I am testing with the simulator now. There is a bug in driver.c that leads to extra step pulses beeing output on direction changes, I'll run some longer tests with the current PIO code before I check with your version.

**Reply by terjeio** [reply]
Removing this line
https://github.com/grblHAL/RP2040/blob/2f62cdeb8b1789e6f36579e45a5a8e73e10d37c0/driver.c#L546
fixes the extra step pulses problem.

There is a remaining "bug" - setting the step pulse length and delay does work as I did not add the code required. To be fixed in the next build.

I ran a few tests with a 9000 line program without the step count going bad with the current PIO code, running a 10 hour test now. If this does not fail then the shift register timing is (close to) ok after all? Set-up time is 3ns according to my datasheet, a one (or two) cycle delay should be enough to get it well inside?

**Reply by andrewmarles** [reply]
I agree a shorter delay is probably sufficient, I just set it to max because I am conservative and don't need to run it at full speed for my slow milling machine.  Maybe at some future point we can find a way to make it easier to configure at least at compile time - I know that for my machines I would set it up for the slowest timing that meets the performance requirement.  But if you are running a laser or integrated drivers you would want higher step rates.  I will try to get a board running with your fix and the tighter timing on my mill in the next few days and we can see if it seems reliable in that setting.  Thanks for looking at this.

I had not come across something like your simulator board before, very handy tool.

**Comment by andrewmarles** [comment]
So I tried these changes on my milling machine this morning with the side set delays and the change to driver.c noted above.  It is much better, I can jog in all directions now, but it appears that the machine is not getting consistent direction changes.  I need to dig deeper to describe exactly what is missing, but as you run a program the x axis is missing direction changes and starts wandering off to one side.  Also, I need to set the step invert on Y to get any movement at all on the X axis, which is clearly weird.  This is even with the side-step delay set to maximum, so I don't think it is any kind of setup/hold problem.  I think that the bit signals still aren't quite lining up correctly.  I think it happens more when X and Y are moving together as I get an error 8 when homing if XY home together - the X axis doesn't pull pack correctly after the first search pass.  But If I home the axes individually it is successful.

I am going to switch to a different setup with a smaller machine (at my desk) and also take another board and set it up on a logic analyzer so that I can dig into this more and help understand exactly where things are going wrong, at the very least I want to understand if it is something specific to my board.

**Reply by terjeio** [reply]
I am seeing issues too - experimenting with PIO code delays and running 10h tests. Step count after a run has been 1-7 steps off, thinking of disconnecting the dir signal and counting the absolute number of step pulses to see if the error is in fact larger. The current run had a 1 step error on the Z-axis for a while that went away.

Strange thing is that adding/changing the delay does not seem to matter much in my setup.

> Also, I need to set the step invert on Y to get any movement at all on the X axis, which is clearly weird.

Weird indeed, I do not see that with my setup.

---

The first instruction (pull block) should have a delay added as well? And perhaps we should try with a reduced PIO clock frequency for the main program?

I also want to run the 10h test with another controller to see if my simulator is behaving.

**Reply by terjeio** [reply]
I am still running tests, now with a Teensy 4.1.

The RP2040 has a step count error of max. 7 steps after running the 10h test - a 186000 line file with relatively small moves in Z. Not much - and I guess it would be hard to detect unless the machine is high precision. With the default 250 steps/mm the accumulated error is max 28 &#181;m. What is a bit worrying is that the count error is not consistent between runs.

Is it the PIO code, grblHAL core code or my machine simulator which is at fault?
To find out I am running tests with other controllers - first out is the Teensy 4.1. After two runs I have a 1 step error in the Y axis and up to 71 steps in Z, X none. Again the result is not consistent between runs.

The RP2040 has a unique software implementation of the float functions, could this be part of it? Teensy 4.1 has a hardware FPU.
I guess the FP math should be ruled in or out as a reason as the precision is not infinite - rounding errors could accumulate?

I plan to run the same test with a 8-bit Mega (with the original code) as well as with both a FPU and non-FPU ARM processor to see if I can narrow down the cause of this. It is going to take some time...

**Reply by andrewmarles** [reply]
I have found that on my dev board with A4988 drivers I can't physically measure a deviation at the end of a program (when running with side-set delays) and using a steps/mm of 800.  So I think that the direction errors that I was seeing on the mill with external drivers may be related to the step/direction timing on the output of the shift register.  I have seen similar issues with external drivers on both LinuxCNC machines (parallel port step/dir) as well as on one of our own machines running GRBLHAL.

However, this particular machine is running well with a Teensy4.1 GRBLHAL board so I will keep digging into this to see if I can better understand what the problem might be on the RP2040 board (if any).

Rounding errors certainly can accumulate across various FPU implementations although these should be standards based so that any errors are consistent.  However, since the RP2040 is a new chip it is entirely possible that things are not 100%, such things have happened before.

**Reply by terjeio** [reply]
I have tweaked your PR so that clock and latch pulses are equal length and tuned them to make each transaction take about 0.8 &#181;s. Minimum dir change to step delay cannot be shorter than this. I will soon commit the changed code.

FYI I have now run some 10h tests with an 8-bit Arduino Mega, this also ends up with a small and random number of erroneous step counts.

**Reply by Volksolive** [reply]
It could be nice to send directly step and dir to the board instead of G-code and bypass the planner / motion-control to see if there is any issue in the PIO code. 
Anyway, it seems like all board have little deviation in all axis .. 
Do you have a way to test your simulator ? Just to be sure that there is nothing there that could explain these errors.

**Reply by terjeio** [reply]
> It could be nice to send directly step and dir to the board instead of G-code and bypass the planner / motion-control to see if there is any issue in the PIO code.

You can do that by setting up a stepper_t struct and calling hal.stepper.pulse_start(), possibly from a small plugin that adds a M-code?
However, I doubt it is the PIO code as all the drivers I have tested with show the same issue.

> Anyway, it seems like all board have little deviation in all axis ..

Even legacy grbl on a Mega2560 shows this behaviour, what bother me a bit is that it is random.

> Do you have a way to test your simulator ? 

I think so, could start by feeding pulse trains to it. If it is a timing problem related to dricetion changes that could be harder to test for.
Currently I have a single sided prototype board with a bit of jump wires, I want to do the testing with a professional made board. I have to order some and get it done.

**Reply by andrewmarles** [reply]
With your CNC simulator, how do you determine your baseline/expected pulsetrain?

**Reply by terjeio** [reply]
@andrewmarles Currently I have only tested with running long gcode programs that starts from 0,0,0 and ends up back there. This should result in the simulator counters coming back to zero as well. I have run some straight cuts and short programs too to check that the count values are correct - these has not failed.

I have an [Analog Discovery 2](https://digilent.com/reference/test-and-measurement/analog-discovery-2/start) that I believe can generate pulse trains and also count them, I could run some tests with that.
